internal/api/impl/v1/dashboard/validator/validator.go

```diff
@@ -0,0 +1,220 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/cuecontext"
+	"cuelang.org/go/cue/load"
+	"github.com/perses/perses/internal/config"
+	v1 "github.com/perses/perses/pkg/model/api/v1"
+	"github.com/sirupsen/logrus"
+)
+
+/*
+ * A Validator can be used to run checks on Dashboard objects, based on cuelang definitions
+ */
+type Validator interface {
+	Process(dashboard *v1.Dashboard) error
+	LoadSchemas(path string)
+}
+
+type validator struct {
+	schemasConf config.Schemas
+	context     *cue.Context
+	baseDef     cue.Value
+	schemas     *sync.Map
+}
+
+const baseDefFileName = "base.cue"
+const pluginsFolderName = "plugins"
+
+/*
+ * Instanciate a validator
+ */
+func New(schemasConf config.Schemas) Validator {
+	return &validator{
+		schemasConf: schemasConf,
+		context:     cuecontext.New(),
+	}
+}
+
+/*
+ * Initialize the validator
+ * NB: this logic was part of New() before. It was extracted to avoid it being called by unrelated unit tests
+ */
+func (v *validator) initSchemas() {
+	// load the base panel definition
+	baseDefPath := fmt.Sprintf("%s/%s", v.schemasConf.Path, baseDefFileName)
+	data, err := os.ReadFile(baseDefPath)
+	if err != nil {
+		logrus.WithError(err).Errorf("Not able to read the base panel definition file at %s, shutting down..", baseDefPath)
+	}
+	baseDef := v.context.CompileBytes(data)
+
+	// initial load of plugins schemas
+	pluginsPath := fmt.Sprintf("%s/%s", v.schemasConf.Path, pluginsFolderName)
+	schemas, err := loadSchemas(v.context, baseDef, pluginsPath)
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+	}
+
+	v.baseDef = baseDef
+	v.schemas = schemas
+}
+
+/*
+ * Validate a dashboard.
+ * The dashboard's panels are matched against the known list of CUE definitions (schemas).
+ * If no schema matches, the validation fails.
+ */
+func (v *validator) Process(dashboard *v1.Dashboard) error {
+	logrus.Tracef("Dashboard to validate: %+v", dashboard)
+
+	// init the validator if it wasn't already
+	if s, _ := v.baseDef.String(); len(s) == 0 {
+		v.initSchemas()
+	}
+
+	var res error
+
+	// go through the panels list
+	// the processing stops as soon as it detects an invalid panel TODO: can probably be improved
+	for k, panel := range dashboard.Spec.Panels {
+		logrus.Tracef("Panel to validate: %s", string(panel))
+
+		// compile the JSON panel into a CUE Value
+		value := v.context.CompileBytes(panel)
+
+		// retrieve panel's kind
+		kind, err := value.LookupPath(cue.ParsePath("kind")).String()
+		if err != nil {
+			logrus.Error("This panel doesn't contain the required \"kind\" field")
```

Comment:
I think I would tend to put a warning log here instead of an error.
It would have been an error log if the system is in a bad shape let's say. But here it can be totally valid and the code is accurate to invalidate this dashboard. So that's why a `warning` seems more accurate for me.